src/state.c

```diff
@@ -134,3 +136,39 @@ void peer_exchange(client_state *state) {
   free_vec_kv_pair(&known_ports);
   free(message);
 }
+
+void broadcast_want(client_state *state) {
+  vector_kv_pair clients_connected = collect_table(&state->file_descriptors);
+  vector_kv_pair pieces_wanted = collect_table(&state->pieces_want);
+
+  // Craft the ask message
+  size_t ask_message_size =
+    sizeof(ask_message) + (4 * sizeof(uint32_t)) * pieces_wanted.size;
+  ask_message* message = malloc(ask_message_size);
```

Comment:
don't malloc for fixed sized structs

Reply:
and yeah since we have multiple "asks", you would need to craft a new ask message for every piece that you want